Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,242 @@
+        apply pow_le_pow_left (Int.ofNat_nonneg (n*a*B+1))
+        rw [mul_add]
+        simp only [cast_succ, cast_mul, mul_one, add_le_add_iff_left]
+        exact_mod_cast one_le_mul (one_le_of_lt hn) (one_le_natNorm_of_ne_zero A hA)
+      _ < (B + 1) ^ (n - m) * (B + 1) ^ m := by
+        simp only [gt_iff_lt, Int.succ_ofNat_pos, pow_pos, mul_lt_mul_right]
+        have h1 : ((n * a) ^ (m / ((n : ℝ) - m))) ^ ((n : ℝ) - m) <
+            ((B + 1): ℝ) ^ ((n : ℝ) - m) := by -- prove the goal with rpow
+          apply Real.rpow_lt_rpow /- this creates 3 goals: 0 ≤ (↑n * ↑a) ^ (↑m / (↑n - ↑m)),
+                                      (↑n * ↑a) ^ (↑m / (↑n - ↑m)) < ↑B + 1 and 0 < ↑n - ↑m -/
+          · apply rpow_nonneg
+            exact_mod_cast Nat.zero_le (n * a)
+          · rw [← norm_eq_natNorm]
+            exact lt_floor_add_one ((n * ‖A‖) ^ (m / ((n : ℝ ) - ↑m)))
+          · simp only [sub_pos, cast_lt, hn]
+        have h2 : ((n * a) ^ (m / ((n : ℝ) - m))) ^ ((n : ℝ) - m) =
+            ((n : ℝ) * a) ^ (m : ℝ) := by
+          rw [← rpow_mul (by exact_mod_cast Nat.zero_le (n * a)) (m / (n - m)) (n - m),
+            hcompexp m n hn]
+        rw [h2] at h1
+        nth_rw 2 [← Nat.cast_sub (le_of_lt hn)] at h1
+        rw [rpow_natCast ((↑B + 1)) (n - m), rpow_natCast ] at h1
+        exact_mod_cast h1
+      _  = ↑((B + 1) ^ n) := by
+        rw [mul_comm, pow_mul_pow_sub]
+        simp only [cast_pow, cast_add, cast_one]
+        exact (le_of_lt hn)
+
+--end (2)
+
+end preparation
+
+theorem exists_ne_zero_int_vec_norm_le  (hA_nezero : A ≠ 0)  : ∃ (t : Fin n → ℤ), t ≠ 0 ∧
+    A.mulVec t = 0 ∧ ‖t‖ ≤ ((n * ‖A‖)^((m : ℝ ) / (n - m))) := by
+  --Pigeonhole
+  rcases Finset.exists_ne_map_eq_of_card_lt_of_maps_to
+    (card_S_lt_card_T m n A hn hA_nezero) (image_T_subset_S m n A) with ⟨ x, hxT,y, hyT ,hneq, hfeq⟩
+  use x-y
+  -- proof that x - y ≠ 0
+  refine ⟨sub_ne_zero.mpr hneq, ?_, ?_⟩
+  -- x-y is a solution
+  simp only [mulVec_sub, sub_eq_zero, hfeq]
+  ---Inequality
+  have n_mul_norm_A_pow_e_nonneg : 0 ≤ (n * ‖A‖)^e := by
+    rw [norm_eq_natNorm]
+    apply Real.rpow_nonneg
+    exact_mod_cast mul_nonneg (le_of_lt (lt_trans hm hn)) (Nat.zero_le A.natNorm)
+  rw [← norm_col, norm_le_iff n_mul_norm_A_pow_e_nonneg]
+  intro i j
+  rw [Finset.mem_Icc] at hyT
+  rw [Finset.mem_Icc] at hxT
+  simp only [col_apply, Pi.sub_apply, ge_iff_le]
+  rw [Int.norm_eq_abs, ← Int.cast_abs]
+  have : |x i - y i| ≤ ((B' i) : ℝ ) := by
+    rw [Int.cast_abs, Int.cast_sub, abs_le]
+    constructor
+    · simp only [neg_le_sub_iff_le_add]
+      rw [← Int.cast_add, Int.cast_le]
+      apply le_trans (hyT.2 i)
+      simp only [le_add_iff_nonneg_left]
+      exact hxT.1 i
+    · simp only [ tsub_le_iff_right]
+      rw [← Int.cast_add,Int.cast_le]
+      apply le_trans (hxT.2 i)
+      simp only [tsub_le_iff_right, le_add_iff_nonneg_right]
+      exact hyT.1 i
+  exact le_trans this (Nat.floor_le n_mul_norm_A_pow_e_nonneg)
```

Comment:
Can you use `refine ... ?_ ...`, where the `?_` corresponds to `this` here? It would make the proof nicer.

Reply:
Done!